File: handlers/handlers_test.go
package handlers

import (
	"bytes"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"path/filepath"
	"runtime"
	"testing"

	"ant_colony_golang_backend/database"
	"ant_colony_golang_backend/models"
)

func setupTestDB(t *testing.T) (*AntColonyHandler, string) {
	_, filename, _, _ := runtime.Caller(0)
	baseDir := filepath.Dir(filepath.Dir(filename))

	sqlPath := filepath.Join(baseDir, "database", "database.sql")
	db, err := database.InitDB(":memory:", sqlPath)
	if err != nil {
		t.Fatalf("Failed to init in-memory db: %v", err)
	}

	handler := NewAntColonyHandler(db, baseDir)
	return handler, baseDir
}

func TestGetTurbinesMap(t *testing.T) {
	handler, _ := setupTestDB(t)

	req := httptest.NewRequest(http.MethodGet, "/ant-colony/get-turbines-map", nil)
	rr := httptest.NewRecorder()

	handler.GetTurbinesMap(rr, req)

	if rr.Code != http.StatusOK {
		t.Fatalf("Expected 200 OK, got %d", rr.Code)
	}

	var data []models.AntColonyMap
	if err := json.Unmarshal(rr.Body.Bytes(), &data); err != nil {
		t.Fatalf("Failed to decode JSON: %v", err)
	}

	if len(data) == 0 {
		t.Fatalf("Expected non-empty turbines map list")
	}

	if data[0].TurbineID == 0 || data[0].TurbineName == "" {
		t.Fatalf("Invalid turbine map entry: %+v", data[0])
	}
}

func TestGetSubsystems(t *testing.T) {
	handler, _ := setupTestDB(t)

	req := httptest.NewRequest(http.MethodGet, "/ant-colony/get-subsystems", nil)
	rr := httptest.NewRecorder()

	handler.GetSubsystems(rr, req)

	if rr.Code != http.StatusOK {
		t.Fatalf("Expected 200 OK, got %d", rr.Code)
	}

	var data []models.Subsystems
	if err := json.Unmarshal(rr.Body.Bytes(), &data); err != nil {
		t.Fatalf("Failed to decode JSON: %v", err)
	}

	if len(data) == 0 {
		t.Fatalf("Expected non-empty subsystems list")
	}

	if data[0].SubsystemID == 0 || data[0].SubsystemName == "" {
		t.Fatalf("Invalid subsystem entry: %+v", data[0])
	}
}

func TestRunRouteOptimizerAntColony(t *testing.T) {
	handler, _ := setupTestDB(t)

	payload := []models.TurbineFaults{
		{TurbineID: 2, SubsystemName: "Electrical System", FaultType: "Minor"},
		{TurbineID: 3, SubsystemName: "Rotor Hub", FaultType: "Major"},
	}
	body, _ := json.Marshal(payload)

	req := httptest.NewRequest(http.MethodPost, "/ant-colony/run-route-optimizer?algorithm=Ant%20Colony", bytes.NewBuffer(body))
	rr := httptest.NewRecorder()

	handler.RunRouteOptimizer(rr, req)

	if rr.Code != http.StatusOK {
		t.Fatalf("Expected 200 OK, got %d. Body: %s", rr.Code, rr.Body.String())
	}

	var res models.AntColonyPath
	if err := json.Unmarshal(rr.Body.Bytes(), &res); err != nil {
		t.Fatalf("Failed to decode response JSON: %v", err)
	}

	if len(res.TurbineOrder) == 0 || len(res.TurbineOrderToShow) == 0 {
		t.Fatalf("Expected non-empty turbine order")
	}
	if res.BestPathLength <= 0 {
		t.Fatalf("Expected positive best path length")
	}
}

func TestRunRouteOptimizerGenetic(t *testing.T) {
	handler, _ := setupTestDB(t)

	payload := []models.TurbineFaults{
		{TurbineID: 2, SubsystemName: "Electrical System", FaultType: "Minor"},
		{TurbineID: 3, SubsystemName: "Rotor Hub", FaultType: "Major"},
	}
	body, _ := json.Marshal(payload)

	req := httptest.NewRequest(http.MethodPost, "/ant-colony/run-route-optimizer?algorithm=Genetic", bytes.NewBuffer(body))
	rr := httptest.NewRecorder()

	handler.RunRouteOptimizer(rr, req)

	if rr.Code != http.StatusOK {
		t.Fatalf("Expected 200 OK, got %d", rr.Code)
	}

	var res models.AntColonyPath
	if err := json.Unmarshal(rr.Body.Bytes(), &res); err != nil {
		t.Fatalf("Failed to decode response JSON: %v", err)
	}

	if len(res.TurbineOrder) == 0 {
		t.Fatalf("Expected non-empty turbine order")
	}
}

func TestRunRouteOptimizerMemetic(t *testing.T) {
	handler, _ := setupTestDB(t)

	payload := []models.TurbineFaults{
		{TurbineID: 2, SubsystemName: "Electrical System", FaultType: "Minor"},
		{TurbineID: 3, SubsystemName: "Rotor Hub", FaultType: "Major"},
	}
	body, _ := json.Marshal(payload)

	req := httptest.NewRequest(http.MethodPost, "/ant-colony/run-route-optimizer?algorithm=Memetic", bytes.NewBuffer(body))
	rr := httptest.NewRecorder()

	handler.RunRouteOptimizer(rr, req)

	if rr.Code != http.StatusOK {
		t.Fatalf("Expected 200 OK, got %d", rr.Code)
	}

	var res models.AntColonyPath
	if err := json.Unmarshal(rr.Body.Bytes(), &res); err != nil {
		t.Fatalf("Failed to decode response JSON: %v", err)
	}

	if len(res.TurbineOrder) == 0 {
		t.Fatalf("Expected non-empty turbine order")
	}
	if len(res.TurbineOrderToShow) == 0 || res.TurbineOrderToShow[0] != "Doca" || res.TurbineOrderToShow[len(res.TurbineOrderToShow)-1] != "Doca" {
		t.Fatalf("Expected TurbineOrderToShow to start and end with Doca: %v", res.TurbineOrderToShow)
	}
}

func TestRunRouteOptimizerSingleTurbine(t *testing.T) {
	handler, _ := setupTestDB(t)

	payload := []models.TurbineFaults{
		{TurbineID: 2, SubsystemName: "Electrical System", FaultType: "Minor"},
	}
	body, _ := json.Marshal(payload)

	req := httptest.NewRequest(http.MethodPost, "/ant-colony/run-route-optimizer?algorithm=Ant%20Colony", bytes.NewBuffer(body))
	rr := httptest.NewRecorder()

	handler.RunRouteOptimizer(rr, req)

	if rr.Code != http.StatusOK {
		t.Fatalf("Expected 200 OK, got %d", rr.Code)
	}

	var res models.AntColonyPath
	if err := json.Unmarshal(rr.Body.Bytes(), &res); err != nil {
		t.Fatalf("Failed to decode response JSON: %v", err)
	}

	if len(res.TurbineOrderToShow) != 3 {
		t.Fatalf("Expected 3 items in TurbineOrderToShow (Doca -> Turbine -> Doca), got %d: %v", len(res.TurbineOrderToShow), res.TurbineOrderToShow)
	}
	if res.TurbineOrderToShow[0] != "Doca" || res.TurbineOrderToShow[2] != "Doca" {
		t.Fatalf("Expected start and end at Doca: %v", res.TurbineOrderToShow)
	}
}

func TestRunRouteOptimizerEmptyPayload(t *testing.T) {
	handler, _ := setupTestDB(t)

	body, _ := json.Marshal([]models.TurbineFaults{})
	req := httptest.NewRequest(http.MethodPost, "/ant-colony/run-route-optimizer?algorithm=Genetic", bytes.NewBuffer(body))
	rr := httptest.NewRecorder()

	handler.RunRouteOptimizer(rr, req)

	if rr.Code != http.StatusOK {
		t.Fatalf("Expected 200 OK, got %d", rr.Code)
	}

	var res models.AntColonyPath
	if err := json.Unmarshal(rr.Body.Bytes(), &res); err != nil {
		t.Fatalf("Failed to decode response JSON: %v", err)
	}

	if len(res.TurbineOrderToShow) == 0 {
		t.Fatalf("Expected non-empty turbine order to show")
	}
	if res.TurbineOrderToShow[0] != "Doca" || res.TurbineOrderToShow[len(res.TurbineOrderToShow)-1] != "Doca" {
		t.Fatalf("Expected start and end with Doca: %v", res.TurbineOrderToShow)
	}
}

File: rust/tests/handlers_test.rs
use std::path::PathBuf;
use std::sync::{Arc, Mutex};
use axum::body::Body;
use axum::http::{Request, StatusCode};
use tower::ServiceExt;

use ant_colony_rust_backend::models::{
    AntColonyMap, AntColonyPath, Subsystems, TurbineFaults,
};
use ant_colony_rust_backend::{create_app, database, AppState};

fn setup_test_app() -> axum::Router {
    let manifest_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    let base_dir = manifest_dir.parent().unwrap().to_path_buf();
    let sql_path = base_dir.join("database").join("database.sql");

    let conn = database::init_db(":memory:", sql_path.to_str().unwrap())
        .expect("Failed to init in-memory DB");

    let state = Arc::new(AppState {
        db: Mutex::new(conn),
        base_dir,
    });

    create_app(state)
}

#[tokio::test]
async fn test_get_turbines_map() {
    let app = setup_test_app();

    let req = Request::builder()
        .uri("/ant-colony/get-turbines-map")
        .body(Body::empty())
        .unwrap();

    let resp = app.oneshot(req).await.unwrap();
    assert_eq!(resp.status(), StatusCode::OK);

    let body = axum::body::to_bytes(resp.into_body(), usize::MAX).await.unwrap();
    let data: Vec<AntColonyMap> = serde_json::from_slice(&body).unwrap();

    assert!(!data.is_empty());
    assert_ne!(data[0].turbine_id, 0);
    assert!(!data[0].turbine_name.is_empty());
}

#[tokio::test]
async fn test_get_subsystems() {
    let app = setup_test_app();

    let req = Request::builder()
        .uri("/ant-colony/get-subsystems")
        .body(Body::empty())
        .unwrap();

    let resp = app.oneshot(req).await.unwrap();
    assert_eq!(resp.status(), StatusCode::OK);

    let body = axum::body::to_bytes(resp.into_body(), usize::MAX).await.unwrap();
    let data: Vec<Subsystems> = serde_json::from_slice(&body).unwrap();

    assert!(!data.is_empty());
    assert_ne!(data[0].subsystem_id, 0);
    assert!(!data[0].subsystem_name.is_empty());
}

#[tokio::test]
async fn test_run_route_optimizer_ant_colony() {
    let app = setup_test_app();

    let payload = vec![
        TurbineFaults {
            turbine_id: 2,
            turbine_name: None,
            subsystem_name: "Electrical System".to_string(),
            fault_type: "Minor".to_string(),
        },
        TurbineFaults {
            turbine_id: 3,
            turbine_name: None,
            subsystem_name: "Rotor Hub".to_string(),
            fault_type: "Major".to_string(),
        },
    ];

    let body_bytes = serde_json::to_vec(&payload).unwrap();
    let req = Request::builder()
        .method("POST")
        .uri("/ant-colony/run-route-optimizer?algorithm=Ant%20Colony")
        .header("content-type", "application/json")
        .body(Body::from(body_bytes))
        .unwrap();

    let resp = app.oneshot(req).await.unwrap();
    assert_eq!(resp.status(), StatusCode::OK);

    let body = axum::body::to_bytes(resp.into_body(), usize::MAX).await.unwrap();
    let res: AntColonyPath = serde_json::from_slice(&body).unwrap();

    assert!(!res.turbine_order.is_empty());
    assert!(!res.turbine_order_to_show.is_empty());
    assert!(res.best_path_length > 0.0);
}

#[tokio::test]
async fn test_run_route_optimizer_genetic() {
    let app = setup_test_app();

    let payload = vec![
        TurbineFaults {
            turbine_id: 2,
            turbine_name: None,
            subsystem_name: "Electrical System".to_string(),
            fault_type: "Minor".to_string(),
        },
        TurbineFaults {
            turbine_id: 3,
            turbine_name: None,
            subsystem_name: "Rotor Hub".to_string(),
            fault_type: "Major".to_string(),
        },
    ];

    let body_bytes = serde_json::to_vec(&payload).unwrap();
    let req = Request::builder()
        .method("POST")
        .uri("/ant-colony/run-route-optimizer?algorithm=Genetic")
        .header("content-type", "application/json")
        .body(Body::from(body_bytes))
        .unwrap();

    let resp = app.oneshot(req).await.unwrap();
    assert_eq!(resp.status(), StatusCode::OK);

    let body = axum::body::to_bytes(resp.into_body(), usize::MAX).await.unwrap();
    let res: AntColonyPath = serde_json::from_slice(&body).unwrap();

    assert!(!res.turbine_order.is_empty());
}

#[tokio::test]
async fn test_run_route_optimizer_memetic() {
    let app = setup_test_app();

    let payload = vec![
        TurbineFaults {
            turbine_id: 2,
            turbine_name: None,
            subsystem_name: "Electrical System".to_string(),
            fault_type: "Minor".to_string(),
        },
        TurbineFaults {
            turbine_id: 3,
            turbine_name: None,
            subsystem_name: "Rotor Hub".to_string(),
            fault_type: "Major".to_string(),
        },
    ];

    let body_bytes = serde_json::to_vec(&payload).unwrap();
    let req = Request::builder()
        .method("POST")
        .uri("/ant-colony/run-route-optimizer?algorithm=Memetic")
        .header("content-type", "application/json")
        .body(Body::from(body_bytes))
        .unwrap();

    let resp = app.oneshot(req).await.unwrap();
    assert_eq!(resp.status(), StatusCode::OK);

    let body = axum::body::to_bytes(resp.into_body(), usize::MAX).await.unwrap();
    let res: AntColonyPath = serde_json::from_slice(&body).unwrap();

    assert!(!res.turbine_order.is_empty());
    assert!(!res.turbine_order_to_show.is_empty());
    assert_eq!(res.turbine_order_to_show.first().unwrap(), "Doca");
    assert_eq!(res.turbine_order_to_show.last().unwrap(), "Doca");
}

#[tokio::test]
async fn test_run_route_optimizer_single_turbine() {
    let app = setup_test_app();

    let payload = vec![TurbineFaults {
        turbine_id: 2,
        turbine_name: None,
        subsystem_name: "Electrical System".to_string(),
        fault_type: "Minor".to_string(),
    }];

    let body_bytes = serde_json::to_vec(&payload).unwrap();
    let req = Request::builder()
        .method("POST")
        .uri("/ant-colony/run-route-optimizer?algorithm=Ant%20Colony")
        .header("content-type", "application/json")
        .body(Body::from(body_bytes))
        .unwrap();

    let resp = app.oneshot(req).await.unwrap();
    assert_eq!(resp.status(), StatusCode::OK);

    let body = axum::body::to_bytes(resp.into_body(), usize::MAX).await.unwrap();
    let res: AntColonyPath = serde_json::from_slice(&body).unwrap();

    assert_eq!(res.turbine_order_to_show.len(), 3);
    assert_eq!(res.turbine_order_to_show[0], "Doca");
    assert_eq!(res.turbine_order_to_show[2], "Doca");
}

#[tokio::test]
async fn test_run_route_optimizer_empty_payload() {
    let app = setup_test_app();

    let payload: Vec<TurbineFaults> = vec![];
    let body_bytes = serde_json::to_vec(&payload).unwrap();

    let req = Request::builder()
        .method("POST")
        .uri("/ant-colony/run-route-optimizer?algorithm=Genetic")
        .header("content-type", "application/json")
        .body(Body::from(body_bytes))
        .unwrap();

    let resp = app.oneshot(req).await.unwrap();
    assert_eq!(resp.status(), StatusCode::OK);

    let body = axum::body::to_bytes(resp.into_body(), usize::MAX).await.unwrap();
    let res: AntColonyPath = serde_json::from_slice(&body).unwrap();

    assert!(!res.turbine_order_to_show.is_empty());
    assert_eq!(res.turbine_order_to_show.first().unwrap(), "Doca");
    assert_eq!(res.turbine_order_to_show.last().unwrap(), "Doca");
}
